### PR TITLE
[CBR-365] Make a start on the periodic acid-state plugin

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
## Description

This PR _almost_ delivers CBR-365, by making a start on adding a plugin to create periodic acid-state checkpoints, as well as pruning old ones (which is something that the old wallet did). This is an incomplete PR which can hopefully be picked up by @parsonsmatt when he's back from 🚴 . Here is what needs to be done in order to close of CBR-365:

- [x] Remove `undefined` in the code in favour of an actual DB path (which is currently not threaded through);
- [x] Add a proper `SafeCopy` instance for `module Cardano.Wallet.Kernel.DB.Util.IxSet` , cfr:

```haskell
instance SafeCopy a => SafeCopy (IxSet a) where
  getCopy = error "getCopy for IxSet wrapper"
  putCopy = error "putCopy for IxSet wrapper"
```

Not necessary to do this as part of CBR-365 (Can be done as part of CBR-305, but it will cause the plugin to fail).

- [x] Add a machinery to take a snapshot when we kill the node.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CBR-365

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by adinapoli-iohk** [comment]
@parsonsmatt Alright, I have managed to at least provide you with an hopefully sensible `SafeCopy` instance for the `IxSet`. What needs to be done in that area still is to write a `safecopyRoundtrip` in the `MarshallingSpec`, but got out of time.

Once that is done, the only thing remaining to be done is indeed passing the correct DB filepaths to the plugin and use those for the archive pruning.

**Comment by adinapoli-iohk** [comment]
@parsonsmatt alright, I have added both the `Safecopy` instance and the roundtrip property. Have fun with the rest :)

File: wallet-new/src/Cardano/Wallet/Kernel.hs [hunk 1]
```diff
@@ -64,18 +68,25 @@ data DatabaseMode
...
 -- | Use the default paths on disk. See 'DatabasePaths' for more details.
 useDefaultPaths :: DatabaseMode
-useDefaultPaths = UseFilePath (DatabasePaths Nothing Nothing)
+useDefaultPaths = UseFilePath
+    $ DatabasePaths
+        (defaultAcidStatePath (Proxy @DB))
+        defaultSqlitePath
+
+defaultAcidStatePath :: Typeable st => Proxy st -> FilePath
+defaultAcidStatePath p = "state" </> show (typeRep p)
```
Comment (adinapoli) [inline wallet-new/src/Cardano/Wallet/Kernel.hs]: Interesting default name -- why not use `wallet-db-acid` for consistency?

File: wallet-new/src/Cardano/Wallet/Kernel.hs [hunk 2]
```diff
@@ -64,18 +68,25 @@ data DatabaseMode
...
+
+defaultSqlitePath :: FilePath
+defaultSqlitePath = "./wallet-db-sqlite.sqlite"
```
Comment (adinapoli) [inline wallet-new/src/Cardano/Wallet/Kernel.hs]: Bikeshed -- `.sqlite3` or `.sqlite`? 😉

**Comment by adinapoli-iohk** [comment]
<facepalm for having commented with my personal GH handle once again 🤦‍♂️ )

File: wallet-new/src/Cardano/Wallet/Kernel.hs (hunk 1 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/Kernel.hs]: That's what `acid-state`'s `openLocalState` uses. I'd be happy to switch the default to that; seems more friendly.

File: wallet-new/src/Cardano/Wallet/Kernel.hs (hunk 2 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/Kernel.hs]: `sqlite3` is more descriptive I guess

**Pull request merged**